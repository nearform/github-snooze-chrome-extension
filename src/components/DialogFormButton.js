import React, { useCallback, useState } from 'react'
import * as chrono from 'chrono-node'
import {
  Button,
  TextField,
  Dialog,
  DialogActions,
  DialogContent,
  DialogContentText,
  DialogTitle,
  Autocomplete,
  Typography
} from '@mui/material'

const options = ['in 1 hour', 'tomorrow', 'next week', 'next month']

export default function DialogFormButton({
  label,
  title,
  description,
  placeholder,
  onConfirm,
  disabled
}) {
  const [open, setOpen] = useState(false)
  const [value, setValue] = useState()

  const reset = useCallback(() => {
    setOpen(false)
    setValue()
  }, [])

  const handleClickOpen = () => setOpen(true)
  const handleClose = () => reset()

  const handleConfirm = () => {
    onConfirm(value)
    reset()
  }

  const handleChange = (e, value) => setValue(chrono.parseDate(value))

  return (
    <div>
      <Button
        fullWidth
        disabled={disabled}
        color="secondary"
        variant="contained"
        onClick={handleClickOpen}
      >
        {label}
      </Button>
      <Dialog open={open} onClose={handleClose}>
        <DialogTitle>{title}</DialogTitle>
        <DialogContent>
          <DialogContentText sx={{ mb: 2 }}>{description}</DialogContentText>
          <Autocomplete
            options={options}
            onChange={handleChange}
            freeSolo
            renderInput={params => (
              <TextField {...params} label={placeholder} />
            )}
          />
          {value && (
            <Typography sx={{ mt: 1 }}>
              You will be notified on {value.toLocaleString()}
            </Typography>
          )}
        </DialogContent>
        <DialogActions>
          <Button onClick={handleClose}>Cancel</Button>
          <Button disabled={!value} onClick={handleConfirm}>
            Confirm
          </Button>
        </DialogActions>
      </Dialog>
    </div>
  )
}
