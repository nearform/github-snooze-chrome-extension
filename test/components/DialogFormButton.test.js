import React from 'react'
import { fireEvent } from '@testing-library/react'
import { render } from '../renderer'
import { parseDate } from 'chrono-node'
import DialogFormButton from '../../src/components/DialogFormButton'

const props = {
  label: 'label',
  title: 'title',
  description: 'description',
  onConfirm: jest.fn(),
  placeholder: 'placeholder',
  disabled: false
}

describe('DialogFormButton.js', () => {
  test('shows the proper enabled DialogFormButton', () => {
    const { asFragment } = render(<DialogFormButton {...props} />)

    expect(asFragment()).toMatchSnapshot()
  })

  test('shows the proper disabled DialogFormButton', () => {
    const { asFragment } = render(
      <DialogFormButton {...props} disabled={true} />
    )

    expect(asFragment()).toMatchSnapshot()
  })

  test('autocomplete component to support custom snooze intervals', async () => {
    const { findByText, getByRole } = render(<DialogFormButton {...props} />)
    const snoozeMinutes = 27
    const chronoInput = `in ${snoozeMinutes} minutes`

    const openDialogButton = await findByText(props.label)
    fireEvent.click(openDialogButton)

    const input = await getByRole('combobox')
    fireEvent.change(input, {
      target: { value: chronoInput }
    })

    const expectedScheduleMessage = `You will be notified on ${parseDate(
      chronoInput
    ).toLocaleString()}`

    await findByText(expectedScheduleMessage)
  })
})
