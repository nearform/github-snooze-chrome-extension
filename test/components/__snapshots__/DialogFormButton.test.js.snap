// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DialogFormButton.js shows the proper disabled DialogFormButton 1`] = `
<DocumentFragment>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium Mui-disabled MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1g1w32r-MuiButtonBase-root-MuiButton-root"
      disabled=""
      tabindex="-1"
      type="button"
    >
      label
    </button>
  </div>
</DocumentFragment>
`;

exports[`DialogFormButton.js shows the proper enabled DialogFormButton 1`] = `
<DocumentFragment>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1g1w32r-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      label
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
